### PR TITLE
ocamltest: add support for hooks

### PR DESCRIPTION
**Pull request by shindere**
This PR refactorizes ocamltest a bit and adds support for hooks.

A hook lets a test override an action with a shell script.

To illustrate this, the PR also contains a test that has been migrated
and makes use of this new feature.

The PR also adds the -e option to ocamltest's command line to make it
log to stderr rather than to a log file.

### CODE REVIEW

File: testsuite/tests/lib-unix/unix-execvpe/exec.ml [hunk 1]
```diff
@@ -1,3 +1,11 @@
+(* TEST
+   include unix
+   script = "${test_source_directory}/has-execvpe"
+   * script
+   ** bytecode
+   ** native
```
Comment (gasche) [inline testsuite/tests/lib-unix/unix-execvpe/exec.ml]: This is a tree of tests. `script` runs first, and its children `bytecode` and `native` are run if it passes. Note that the magical return code 125 means to skip, not fail.

File: testsuite/tests/lib-unix/unix-execvpe/has-execvpe [hunk 2]
```diff
@@ -0,0 +1,9 @@
+#!/bin/sh
+
+# This script is related to the 'exec.ml' test.
+# It tests whether the OS implements execvpe or not.
+# It makes sense to run the tests only if execvpe is nt implemented.
+# If it is implemented, the test is skipped.
+
+if grep -q HAS_EXECVPE ${ocamlsrcdir}/byterun/caml/s.h; then exit 125; fi
```
Comment (gasche) [inline testsuite/tests/lib-unix/unix-execvpe/has-execvpe]: Is there a way to use a variable name for the exit codes, such as `$EXIT_SUCCESS`, `$EXIT_FAILURE` and `$EXIT_SKIP`? Maybe the driver could always set these variables in the environment when running scripts?

**Comment by shindere** [comment]
Gabriel Scherer (2017/11/16 09:31 +0000):
> gasche commented on this pull request.
> 
> 
> 
> > @@ -1,3 +1,11 @@
> +(* TEST
> +   include unix
> +   script = "${test_source_directory}/has-execvpe"
> +   * script
> +   ** bytecode
> +   ** native
> 
> This is a tree of tests. `script` runs first, and its children `bytecode` and `native` are run if it passes. Note that the magical return code 125 means to skip, not fail.
> 
> > @@ -0,0 +1,9 @@
> +#!/bin/sh
> +
> +# This script is related to the 'exec.ml' test.
> +# It tests whether the OS implements execvpe or not.
> +# It makes sense to run the tests only if execvpe is nt implemented.
> +# If it is implemented, the test is skipped.
> +
> +if grep -q HAS_EXECVPE ${ocamlsrcdir}/byterun/caml/s.h; then exit 125; fi
> 
> Is there a way to use a variable name for the exit codes, such as
> `$EXIT_SUCCESS`, `$EXIT_FAILURE` and `$EXIT_SKIP`? Maybe the driver
> could always set these variables in the environment when running
> scripts?

Sure, will do that, nice suggestion, thanks.

Re: the variable names, how about TEST_PASS, TEST_SKIP and TEST_FAIL?

**Comment by gasche** [comment]
Yes, your names are fine as well -- in particular they don't look identical to C library conventions for exit code naming that are used for subtly different purposes.

**Comment by shindere** [comment]
@dra27 @nojb :

Currently the lib-unix/unix-execvpe test fails with the lgo below.

Given the shape of the error message I am suspecting a string conversion
problem but can't find it.

Is one of you guys able to figure out what's wrong?

Otherwise I am thinking about splitting this PR in two: one for ocamltest
and one for the migration of this test so that ocamltest canbe merged.

Log:

...
Running action 1/1 (script)
Commandline: C:\cygwin\home\shindere\src\ocaml\testsuite\tests\lib-unix\unix-execvpe/has-execvpe exec.ml
run_win32.c:323: CreateProcess failed: Â숀帰Ȯ栐ȯ
Action 1/1 (script) failed for the following reason:
Running script C:\cygwin\home\shindere\src\ocaml\testsuite\tests\lib-unix\unix-execvpe/has-execvpe with arguments exec.ml: command
C:\cygwin\home\shindere\src\ocaml\testsuite\tests\lib-unix\unix-execvpe/has-execvpe exec.ml
failed with exit code -1

**Comment by shindere** [comment]
@octachron : by the way this PR includes a "compare_programs" variable which
is taken into account by both the bytecode and the native tests. The
comparison is enabled by default and the variable should be set to exactly
"false" to disable it.

**Comment by nojb** [comment]
I looked a bit into the AppVeyor failure: it seems you are trying to execute a `#!` script with the native call `CreateProcess`.  This won't work. You can try calling `sh` directly, but this already shows that using shell scripts is inherently problematic.

**Comment by shindere** [comment]
Nicolás Ojeda Bär (2017/11/17 12:43 +0000):
> I looked a bit into the AppVeyor failure: it seems you are trying to
> execute a `#!` script with the native call `CreateProcess`.  This
> won't work. You can try calling `sh` directly, but this already shows
> that using shell scripts is inherently problematic.

Ah of course, well spotted. Do you have an alternative solution in mind?

**Comment by xavierleroy** [comment]
One solution would be to make an explicit invocation of `sh`:
```
script = "sh"
script_argument1 = "${test_source_directory}/has-execvpe"
```
I'm making up syntax here.  A nicer syntax would give the command and the arguments simultaneously:
```
script = ["sh"; "${test_source_directory}/has-execvpe"]
```
Then, if you so wish you could get rid of the script and inline it:
```
script = ["sh"; "-c"; "if blablabla; then exit 125; else exit 0; fi" ]
```

**Comment by shindere** [comment]
Xavier Leroy (2017/11/17 14:21 +0000):
> One solution would be to make an explicit invocation of `sh`:
> ```
> script = "sh"
> script_argument1 = "${test_source_directory}/has-execvpe"
> ```
> I'm making up syntax here.  A nicer syntax would give the command and the arguments simultaneously:
> ```
> script = ["sh"; "${test_source_directory}/has-execvpe"]
> ```
> Then, if you so wish you could get rid of the script and inline it:
> ```
> script = ["sh"; "-c"; "if blablabla; then exit 125; else exit 0; fi" ]
> ```

Ah that's a nice one, thanks!

I was thinking that the script test could call ${SHELL} by default, so
you would write either

script = "scriptfile"

or

script = "-c command"

and both would work.

The advantages I can see are a lighter syntax and also that it avoids
the risk that the user forgets to mention the shell.

The drawback, IMO, would be that this solution is less general in that
it does not let a test specify an alternative interpreter, except by
overwriting the SHELL environment variable.

Any thought @xavierleroy ? Others?

**Comment by nojb** [comment]
Personally I prefer to be as explicit as possible, so would be in favor of writing the `sh` invocation and not assuming it.

For example there are tests that use OCaml scripts to decide whether to run and these could be invoked directly without going through `sh`.

**Comment by shindere** [comment]
Nicolás Ojeda Bär (2017/11/17 14:38 +0000):
> Personally I prefer to be as explicit as possible, so would be in favor of writing the `sh` invocation and not assuming it.
> 
> For example there are tests that use OCaml scripts to decide whether
> to run and these could be invoked directly without going through `sh`.

OK will do that then. Thanks!!

**Comment by dra27** [comment]
Bear in mind if you go down the route of embedding shell calls like that, we'll hit the same problems as ocamlbuild used to with quoting at some point. Even doing the arguments as a list, you'll need the whole of https://github.com/dra27/opam/blob/windows/src/core/opamProcess.ml#L24-L202 in order to be sure that you're directly calling a Cygwin process without some escaping nonsense getting in the way. I would suggest if it's to be done with scripts for now, then it should be a simple "driver command" (sh, ocaml, bash, etc.) and a single filename, though I of course agree with @xavierleroy that it would be nicer not to be forced to spread these things over multiple files.

It's hard to make this pronouncement at this stage before most/all the tests are converted, but I would hope that we could ultimately move towards replacing the shell scripts with extremely simple OCaml scripts, possibly with a small utilities module.

**Comment by shindere** [comment]
David Allsopp (2017/11/17 17:28 +0000):
> Bear in mind if you go down the route of embedding shell calls like
> that, we'll hit the same problems as ocamlbuild used to with quoting
> at some point.

Indeed, fully agreed. I understand how a self-contained approach is
appealing, in principle, but I also believe it has the practical
limitations David points out.

When you look at it, that would mean embedding a language (shell
scripting language) into a DSL which is itself embedded in an OCaml
file. Frm that persective, it does not look that exciting to me.

Also, it would mean either allowing rather long lines (if you write
something like

(* TEST
  * script = "sh -c ..."
  * script
*)
you have less than 80 characters left for your command, that's really
not much)

or it would mean allowing multi-line strings in the ocamltest DSL.

And also, even if the external files based solution may seem a bit
heavier at first glance, I believe it will be more sustainable in the
long run and will also lend itself more to sharing scripts between
different tests than the embedded way.

So this is all to say I fully agree with you, @dra27 :)

> It's hard to make this pronouncement at this stage before most/all the
> tests are converted,

Agreed, too.

> but I would hope that we could ultimately move
> towards replacing the shell scripts with extremely simple OCaml
> scripts, possibly with a small utilities module.

I think the ocamltest/builtin_actions module is what you are looking
for.

**Comment by shindere** [comment]
Gabriel Scherer (2017/11/16 09:31 +0000):
> Is there a way to use a variable name for the exit codes, such as
> `$EXIT_SUCCESS`, `$EXIT_FAILURE` and `$EXIT_SKIP`? Maybe the driver
> could always set these variables in the environment when running
> scripts?

So, thanks again for this pretty cool suggestion, @gasche.

The current version of the PR (which is not yet completely ready to be
merged, btw) implements it with the names I proposed the other day:
TEST_PASS, TEST_SKIP and TEST_FAIL.

One thing I'd like to fix before this gets merged is that the error
messages from the C code are not printed correctly under Windows, as can
be seen from the log excerpt posted above.

I compared the code that deals with formatting error messages
(report_error function in ocamltest/run_win32.c) with the code used in
e.g. byterun/win32.c to do the same thing and the two do not seem
fundamentally different, at least in the way they deal with strings.

Can any Windows expert (@dra27, @nojb, @alainfrisch) understand what's
wrong in ocamltest on this topic?

**Comment by xavierleroy** [comment]
@dra27: yes, shell quoting problems might bite again, a separate shell script is always safer.  Speaking of shell quoting, see also the recent [MPR7672](https://caml.inria.fr/mantis/view.php?id=7672) and feel free to add references to relevant discussions there.

**Comment by shindere** [comment]
So the current version of the PR works, but hooks seem to be not working on
Windows at the moment.

More precisely, when a hook (which is a bash script) wants to call the
program that has just been compiled, it fails to do so. Bash reports the
program is not found. Am I right to assume that this is because the
program's path uses the Windows format?

To reproduce the problem, it is possible to put a ! in front of the grep
command in the has-execvpe script.

So the PR could be merged in its current state, because the migrated
test does not invoke hooks on Windows. It just invokes the has-execvpe
script which tells to skip the remainder of the test and this
(calling scripts) works. It is also possible to wait until this
(hopefully last) problem is fixed. 

Any suggestion?

**Comment by xavierleroy** [comment]
Concerning the possible issue with the directory path, you could run the script with `sh -x` instead of `sh` and see what's in the log.

At any rate, ocamltest would do all of us a great favor by changing working directory itself before running scripts.  Scripts expect to be run from the directory that contains the test.  Asking scripts to chdir themselves or to use absolute path names is probably too much to ask from the authors of the scripts.

**Comment by shindere** [comment]
Xavier Leroy (2017/11/19 15:30 +0000):
> At any rate, ocamltest would do all of us a great favor by changing
> working directory itself before running scripts.

It does so. Hooks are run in the directory where the test is built.

> Scripts expect to be run from the directory that contains the test.

Tjat's what happens, yes.

> Asking scripts to chdir themselves or to use absolute path names is
> probably too much to ask from the authors of the scripts.

It's not really asked from them. It remains true that some of the paths
they are given are absolute.

Generally speaking, this is because the concept of build directory is
not relevant in all situations. While comparing programs, for instance,
there is not one single build directory. Just two programs to compare.

At the moment I am able to make things work by using cygpath in the hook
when it detects that it is run on Windows.

I can't think of a better alternative right now but find the situation
not very satisfying. So if there is a suggestion about a way to
normalize paths from OCaml so that they make bash happy, I'd be more
than interested in hearing about it.

>

**Comment by xavierleroy** [comment]
Thanks for the clarification.  So, the issue is not the working directory, but still it is related to the use of absolute file paths.  

> a way to normalize paths from OCaml so that they make bash happy

Call me obsessive, but what about making all file paths relative to the directory in which the command is run?

**Comment by shindere** [comment]
Xavier Leroy (2017/11/19 16:48 +0000):
> Call me obsessive, but what about making all file paths relative to
> the directory in which the command is run?

I fully agree this is a desirable goal. There are two obstacles to
overcome, though:

1. So far we don't copy the reference files for program and compiler
outputs to the test build directories (one per compiler) so it seems
there will always be absolute paths for them.

2. As I tried to explain, one single file may appear in different
contexts. In terms of commands, this means that a filename may appear in
several commands that are run from different working directories. I am
not sure how to do to recompute the right filenames each time, but if
there is a solution to this problem I am definitely willing to implement
it.

**Comment by alainfrisch** [comment]
> A hook lets a test override an action with a shell script.

Why not go directly for OCaml scripts instead?  Either as external files, or directly embedded in the TEST description.

**Comment by shindere** [comment]
Alain Frisch (2017/11/20 08:12 +0000):
> > A hook lets a test override an action with a shell script.
> 
> Why not go directly for OCaml scripts instead?  Either as external
> files, or directly embedded in the TEST description.

Mostly because shell scripts is what we have currantly so it will be
easier to migrate shell scripts to shell scripts than shell scripts to
ocaml scripts.

Also, there are many things that would not be that easy to do from
within an OCaml script that shall not have any dependency other than the
compiler.

**Comment by shindere** [comment]
CI seems happy now. Travis still pending but perhaps hung, I'm not sure.
This PR went through Inria's precheck successfully.

I think this can be merged now.

**Comment by xavierleroy** [comment]
Let's merge and go forth with the migration!

**Comment by shindere** [comment]
Thanks a lot!

**Pull request merged**